File: tui/src/scala/tui/widgets/canvas/CharGrid.scala
package tui.widgets.canvas

import tui.internal.ranges
import tui.{Color, Point}

case class CharGrid(
    width: Int,
    height: Int,
    cells: Array[Char],
    colors: Array[Color],
    cell_char: Char
) extends Grid {

  override def resolution: Point =
    Point(this.width.toDouble - 1.0, this.height.toDouble - 1.0)

  override def save(): Layer =
    Layer(
      string = new String(this.cells),
      colors = this.colors.map(identity)
    )

  override def reset(): Unit = {
    ranges.range(0, this.cells.length) { i =>
      this.cells(i) = ' '
    }
    ranges.range(0, this.colors.length) { i =>
      this.colors(i) = Color.Reset
    }
  }

  override def paint(x: Int, y: Int, color: Color): Unit = {
    val index = y * this.width + x
    if (index < this.cells.length) {
      this.cells(index) = this.cell_char
    }
    if (index < this.colors.length) {
      this.colors(index) = color
    }
  }
}
object CharGrid {
  def apply(width: Int, height: Int, cell_char: Char): CharGrid = {
    val length = width * height
    CharGrid(
      width,
      height,
      cells = Array.fill(length)(' '),
      colors = Array.fill(length)(Color.Reset),
      cell_char
    )
  }
}

File: tui/src/java/tui/widgets/canvas/CharGrid.java
package tui.widgets.canvas;

import tui.Color;
import tui.Point;
import tui.internal.Ranges;

public final class CharGrid implements Grid {
  public final int width;
  public final int height;
  public final char[] cells;
  public final Color[] colors;
  public final char cellChar;

  public CharGrid(int width, int height, char[] cells, Color[] colors, char cellChar) {
    this.width = width;
    this.height = height;
    this.cells = cells;
    this.colors = colors;
    this.cellChar = cellChar;
  }

  public static CharGrid create(int width, int height, char cellChar) {
    int length = width * height;
    char[] cells = new char[length];
    Color[] colors = new Color[length];
    for (int i = 0; i < length; i++) {
      cells[i] = ' ';
      colors[i] = Color.Reset;
    }
    return new CharGrid(width, height, cells, colors, cellChar);
  }

  @Override
  public int width() {
    return width;
  }

  @Override
  public int height() {
    return height;
  }

  @Override
  public Point resolution() {
    return new Point((double) width - 1.0, (double) height - 1.0);
  }

  @Override
  public Layer save() {
    return new Layer(new String(cells), colors.clone());
  }

  @Override
  public void reset() {
    Ranges.range(0, cells.length, i -> cells[i] = ' ');
    Ranges.range(0, colors.length, i -> colors[i] = Color.Reset);
  }

  @Override
  public void paint(int x, int y, Color color) {
    int index = y * width + x;
    if (index < cells.length) {
      cells[index] = cellChar;
    }
    if (index < colors.length) {
      colors[index] = color;
    }
  }
}
